src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java

```diff
@@ -0,0 +1,162 @@
+    VirtualFile vf = LocalFileSystem.getInstance().findFileByIoFile(modulePath.toFile());
+    checkPsiFile(PsiManager.getInstance(project).findFile(vf));
+  }
+
+  @Override
+  public boolean registerListener() {
+    EditorEventMulticaster multicaster = EditorFactory.getInstance().getEventMulticaster();
+    disposable = Disposer.newDisposable();
+    multicaster.addDocumentListener(this, disposable);
+    //make sure the correct class file is open?
+    return false;
+  }
+
+  @Override
+  public void unregisterListener() {
+    EditorEventMulticaster multicaster = EditorFactory.getInstance().getEventMulticaster();
+    multicaster.removeDocumentListener(this);
+    disposable.dispose();
+  }
+
+  @Override
+  public void documentChanged(@NotNull DocumentEvent event) {
+    PsiDocumentManager.getInstance(project).commitDocument(event.getDocument());
+    PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(event.getDocument());
+    checkPsiFile(psiFile);
+  }
+
+  private void checkPsiFile(PsiFile psiFile) {
+    if (psiFile != null) {
+      psiFile.accept(new ScalaRecursiveElementVisitor() {
+        @Override
+        public void visitScalaElement(ScalaPsiElement element) {
+          super.visitScalaElement(element);
+          if ((element instanceof ScClass && checkScClass((ScClass) element))) {
+            PsiElement[] children = element.getChildren();
+            Optional<PsiElement> constructor = Arrays.stream(children).filter(
+                ScPrimaryConstructor.class::isInstance).findFirst();
+            if (constructor.isPresent()
+                    && checkScPrimaryConstructor((ScPrimaryConstructor) constructor.get())) {
+              Optional<PsiElement> extendsBlock = Arrays.stream(children).filter(
+                  ScExtendsBlockImpl.class::isInstance).findFirst();
+              if (extendsBlock.isPresent()
+                      && checkExtendsBlock((ScExtendsBlockImpl) extendsBlock.get())) {
+                ApplicationManager.getApplication().invokeLater(callback::callback);
+              }
+            }
+          }
+        }
+      });
+    }
+  }
+
+  private boolean checkScClass(ScClass element) {
+    return className.equals(element.getName());
+  }
+
+  private boolean checkScPrimaryConstructor(ScPrimaryConstructor element) {
+    Map<String, String> args = new HashMap<>();
+    for (String a: arguments) {
+      StringTokenizer tokenizer = new StringTokenizer(a, ":", false);
+      String key = tokenizer.nextToken();
+      String value = tokenizer.nextToken();
+      args.put(key, value);
+    }
+    boolean correct = true;
+    PsiParameter[] parameters = element.parameterList().getParameters();
+    for (PsiParameter param: parameters) {
+      if (!args.containsKey(param.getName())
+          || !args.get(param.getName()).equals(param.getType().getPresentableText())) {
+        correct = false;
+      }
+    }
+    return correct;
```

Comment:
Return early from the loop by returning false to reduce iterations and remove the unnecessary `correct` variable